File: backend/src/services/senderOrchestrator.js
/**
 * Sender Orchestrator
 *
 * Called when the "sender" agent type is triggered.
 * Sends all approved emails, updates the agent run record.
 */

import prisma from '../lib/prisma.js';
import { sendApprovedEmails } from './emailSender.js';

export async function runSender(runId) {
  try {
    await prisma.agentRun.update({
      where: { id: runId },
      data: { status: 'running' },
    });

    const result = await sendApprovedEmails();

    await prisma.agentRun.update({
      where: { id: runId },
      data: {
        status: 'completed',
        completedAt: new Date(),
        emailsSent: result.sent,
        errors: result.errors || [],
      },
    });

    console.log(`Sender: done. ${result.sent} emails sent.`);
  } catch (err) {
    console.error('Sender orchestrator error:', err);
    await prisma.agentRun.update({
      where: { id: runId },
      data: {
        status: 'failed',
        completedAt: new Date(),
        errors: [err.message],
      },
    });
  }
}

File: app/services/sender_orchestrator.py
"""
Sender Orchestrator.

Wraps send_approved_emails() with agent run tracking.
"""

from datetime import datetime

from app.database import AsyncSessionLocal
from app.models import AgentRun
from app.services.email_sender import send_approved_emails


async def run_sender(run_id: str):
    async with AsyncSessionLocal() as db:
        try:
            run = await db.get(AgentRun, run_id)
            if run:
                run.status = "running"
                await db.commit()

            result = await send_approved_emails()

            run = await db.get(AgentRun, run_id)
            if run:
                run.status = "completed"
                run.completed_at = datetime.utcnow()
                run.emails_sent = result.get("sent", 0)
                run.errors = result.get("errors", [])
                await db.commit()

            print(f"Sender: done. {result.get('sent', 0)} emails sent.")

        except Exception as e:
            print(f"Sender orchestrator error: {e}")
            run = await db.get(AgentRun, run_id)
            if run:
                run.status = "failed"
                run.completed_at = datetime.utcnow()
                run.errors = [str(e)]
                await db.commit()
